inference/src/translator/annotation.cpp

```diff
@@ -34,8 +34,83 @@ void AnnotatedText::appendSentence(string_view prefix, std::vector<string_view>:
   annotation.token_begin_.push_back(offset);
 }
 
-void AnnotatedText::handleEndingWhitespace(string_view whitespace) {
-  text.append(whitespace.data(), whitespace.size());
+bool AnnotatedText::shouldOmitSpaceBetweenSentences() const {
+  if (targetLanguage_.empty()) {
+    // The target language is not specified, so we should not make assumptions about
+    // whether or not the language's script should omit whitespace.
+    return false;
+  }
+
+  return (
+    // Japanese does not use space between sentences.
+    targetLanguage_ == "ja" ||
+    // Korean does not use space between sentences.
+    targetLanguage_ == "ko" ||
+    // Chinese does not use space between sentences.
+    targetLanguage_ == "zh"
+  );
+}
+
+bool AnnotatedText::shouldEnsureSpaceBetweenSentences() const {
+  if (targetLanguage_.empty()) {
+    // The target language is not specified, so we should not make assumptions about
+    // whether or not the language's script should omit whitespace.
+    return false;
+  }
+
+  return !shouldOmitSpaceBetweenSentences();
+}
+
+void AnnotatedText::maybeAppendHTMLTagsFromWhitespace(string_view whitespace) {
+  size_t currentIndex = 0;
+  while (currentIndex < whitespace.size()) {
+    // Find the next open bracket '<' for an HTML tag.
+    size_t tagStart = whitespace.find('<', currentIndex);
```

Comment:
Thought: So this can be more complicated, and can break with how the web actually works. `<pre>x < 5</pre>` will render perfectly fine on the web. However, I think this may be matching existing behavior with the HTML parsing.

Suggestion: It may be worth doing a little bit of investigating here.

Reply:
You're definitely right about an edge case like `<pre>x < 5</pre>` being an issue in HTML parsing in general. 

However, at this stage in the algorithm, we are only dealing with pre-parsed HTML tags that were are adjacent to sentence-separating whitespace and need to be preserved. Consider the following example:

I've added the following to this function:
```c++
  std::cout << "!!! Whitespace Value: " << whitespace << std::endl;
```

Then I've run this test case:
```javascript
"<b>The closing for this bold tag and <em>emphasis tag will show up because they are adjacent to a whitespace sentence boundary.</em></b> However, this <u>underline</u> tag will not be present in the middle of a sentence.",
```

And the resulting logs:
```
!!! Whitespace Value: 
!!! Whitespace Value:  
!!! Whitespace Value: 
!!! Whitespace Value: 
!!! Whitespace Value: </em></b> 
!!! Whitespace Value: 
!!! Whitespace Value: 
!!! Whitespace Value: </em></b>
!!! Whitespace Value: 
```

---

I will try to make the comments and documentation more clear here that we are only extracting pre-parsed tags that are adjacent to sentence-separating whitespace.